File: 07_linked_list/07_04_linked_list_cycle/linked_list_cycle.cpp
#include <iostream>

class ListNode {
public:
    int val;
    ListNode* next;

    ListNode(int value = 0, ListNode* nextNode = nullptr) : val(value), next(nextNode) {}
};

bool hasCycle(ListNode* head) {
    ListNode* slow = head;
    ListNode* fast = head;
    while (fast != nullptr && fast->next != nullptr) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) {
            return true;
        }
    }
    return false;
}

int main() {
    // Test Case 1: Cycle present
    ListNode* head1 = new ListNode(1);
    ListNode* node2 = new ListNode(2);
    ListNode* node3 = new ListNode(3);
    head1->next = node2;
    node2->next = node3;
    node3->next = node2;  // Creating a cycle
    std::cout << "Test Case 1: " << (hasCycle(head1) ? "true" : "false") << std::endl;  // Output: true

    // Test Case 2: No cycle
    ListNode* head2 = new ListNode(1, new ListNode(2, new ListNode(3)));
    std::cout << "Test Case 2: " << (hasCycle(head2) ? "true" : "false") << std::endl;  // Output: false

    // Test Case 3: Single node with a cycle
    ListNode* head3 = new ListNode(1);
    head3->next = head3;  // Creating a cycle
    std::cout << "Test Case 3: " << (hasCycle(head3) ? "true" : "false") << std::endl;  // Output: true

    // Test Case 4: Single node without a cycle
    ListNode* head4 = new ListNode(1);
    std::cout << "Test Case 4: " << (hasCycle(head4) ? "true" : "false") << std::endl;  // Output: false

    // Test Case 5: Empty list
    ListNode* head5 = nullptr;
    std::cout << "Test Case 5: " << (hasCycle(head5) ? "true" : "false") << std::endl;  // Output: false

    // Remember to free allocated memory
    delete head1;
    delete head2;
    delete head3;
    delete head4;

    return 0;
}

File: 07_linked_list/07_04_linked_list_cycle/linked_list_cycle.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def has_cycle(head):
    slow = head
    fast = head
    while fast is not None and fast.next is not None:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return True
    return False

# Test Case 1: Cycle present
head1 = ListNode(1)
node2 = ListNode(2)
node3 = ListNode(3)
head1.next = node2
node2.next = node3
node3.next = node2  # Creating a cycle
print("Test Case 1:", has_cycle(head1))  # Output: True

# Test Case 2: No cycle
head2 = ListNode(1, ListNode(2, ListNode(3)))
print("Test Case 2:", has_cycle(head2))  # Output: False

# Test Case 3: Single node with a cycle
head3 = ListNode(1)
head3.next = head3  # Creating a cycle
print("Test Case 3:", has_cycle(head3))  # Output: True

# Test Case 4: Single node without a cycle
head4 = ListNode(1)
print("Test Case 4:", has_cycle(head4))  # Output: False

# Test Case 5: Empty list
head5 = None
print("Test Case 5:", has_cycle(head5))  # Output: False
